### PR TITLE
Prevent out-of-order data being written unless explicitly requested

### PR DESCRIPTION
**Pull request by GemDot**

### CODE REVIEW

File: cpp/arcticdb/version/version_core.cpp
Comment (mehertz) [inline cpp/arcticdb/version/version_core.cpp:542]: The underscores in `_date_range_` won't be rendered.

Comment (mehertz) [inline cpp/arcticdb/version/version_core.cpp:542]: `(SYMBOL)` should be replaced by the actual symbol!

File: python/arcticdb/version_store/library.py
Comment (mehertz) [inline python/arcticdb/version_store/library.py:259]: Needs to be documented in the docstring below

File: python/arcticdb/version_store/library.py
Comment (mehertz) [inline python/arcticdb/version_store/library.py:431]: Needs documenting below. This will apply to every payload member?

File: python/arcticdb/version_store/library.py
Comment (mehertz) [inline python/arcticdb/version_store/library.py:546]: Documentation

File: python/tests/unit/arcticdb/version_store/test_date_range.py
Comment (mehertz) [inline python/tests/unit/arcticdb/version_store/test_date_range.py:79]: Can we also test what happens when you call date range and it's not even a DateTime index please

File: cpp/arcticdb/version/version_core.cpp
Comment (mehertz) [inline cpp/arcticdb/version/version_core.cpp:141]: Remove underscores

Comment (mehertz) [inline cpp/arcticdb/version/version_core.cpp:141]: Error code

File: cpp/arcticdb/version/version_core.cpp
Comment (mehertz) [inline cpp/arcticdb/version/version_core.cpp:109]: Error code

File: python/arcticdb/version_store/_normalization.py
Comment (qc00) [inline python/arcticdb/version_store/_normalization.py:777]: Should really `MultiIndex`

File: python/arcticdb/version_store/_store.py
Comment (qc00) [inline python/arcticdb/version_store/_store.py:488]: Indentation

Comment (qc00) [inline python/arcticdb/version_store/_store.py:488]: Indentation

File: python/arcticdb/version_store/_store.py
Comment (qc00) [inline python/arcticdb/version_store/_store.py:25]: Unused?

File: python/arcticdb/version_store/_store.py
Comment (qc00) [inline python/arcticdb/version_store/_store.py:562]: Consider creating an options object that passes from Python all the way to the `*_impl` in C++.

File: python/arcticdb/version_store/library.py
Comment (qc00) [inline python/arcticdb/version_store/library.py:431]: Why would `write_batch` has a different default from `write`??

File: python/tests/unit/arcticdb/version_store/test_append.py
Comment (qc00) [inline python/tests/unit/arcticdb/version_store/test_append.py:292]: This is not a small change, so really need a lot more tests for each. I would expect:

* `append(ensure_sorted=False)`
* `RangeIndex`
* `MultiIndex`
* Various combinations of ascending, descending and unsorted in both existing data and the argument (you might want to look into Hypothesis)

File: python/tests/unit/arcticdb/version_store/test_append.py
Comment (qc00) [inline python/tests/unit/arcticdb/version_store/test_append.py:318]: The fixture setup is a bit too long, which makes it hard to see what the test is actually checking. (IMO, the code doesn't care about the types of the columns nor the number of rows nor randomness.)

Please consider using existing fixture creation functions.

File: python/tests/unit/arcticdb/version_store/test_date_range.py
Comment (qc00) [inline python/tests/unit/arcticdb/version_store/test_date_range.py:76]: `InternalException` is a bit generic.

For all new errors, a unique error code should be assigned and asserted in tests.

File: python/tests/unit/arcticdb/version_store/test_update.py
Comment (qc00) [inline python/tests/unit/arcticdb/version_store/test_update.py:473]: Do not copy and paste these.

File: python/arcticdb/version_store/_normalization.py
Comment (mehertz) [reply python/arcticdb/version_store/_normalization.py:777]: This is a great point actually - if multiindex is set is `index` not set?

File: python/arcticdb/version_store/library.py
Comment (mehertz) [reply python/arcticdb/version_store/library.py:431]: I think `write` is wrong - it looks to me like it defaults to `False` but it should default to `True` in this API.

Comment (GemDot) [reply python/arcticdb/version_store/library.py:431]: yes, we set it it common to all the cases, isn't it?

File: python/arcticdb/version_store/_normalization.py
Comment (qc00) [reply python/arcticdb/version_store/_normalization.py:777]: `MultiIndex` is just a sub-class of `Index`, but it doesn't directly have `is_monotonic_increasing` as it wouldn't make sense. Will have to drill down to the first level of a `MultiIndex`.

File: python/arcticdb/version_store/library.py
Comment (GemDot) [reply python/arcticdb/version_store/library.py:431]: yes, that should be True as well, thanks

File: python/arcticdb/version_store/_store.py
Comment (GemDot) [reply python/arcticdb/version_store/_store.py:25]: it is used in 2250

File: python/arcticdb/version_store/_normalization.py
Comment (GemDot) [reply python/arcticdb/version_store/_normalization.py:777]: better to discuss with Matt

File: python/tests/unit/arcticdb/version_store/test_date_range.py
Comment (GemDot) [reply python/tests/unit/arcticdb/version_store/test_date_range.py:79]: I added the test, but the check was already on place as internal error. I didn't change the error code in this case, for comp issues

File: python/arcticdb/version_store/_normalization.py
Comment (GemDot) [reply python/arcticdb/version_store/_normalization.py:777]: The logic we have right now already works for multi index. The reason is because _index_to_records function is doing a reset index for the index levels beyond than zero. which basically, converts the index to its single version (relying only in the first level). I added a test for it

File: cpp/arcticdb/version/local_versioned_engine.cpp
Comment (mehertz) [reply cpp/arcticdb/version/local_versioned_engine.cpp:374]: Can we put this in `write_options`? Does that make sense?

**Review by mehertz — APPROVED** [review]

File: python/arcticdb/version_store/_normalization.py
Comment (alexowens90) [inline python/arcticdb/version_store/_normalization.py:868]: I think this should be `UNSORTED` for pickled data?

File: python/tests/conftest.py
Comment (alexowens90) [inline python/tests/conftest.py:253]: Please avoid checking in random formatting/linting changes like this in files that are otherwise untouched

File: python/arcticdb/version_store/_store.py
Comment (alexowens90) [inline python/arcticdb/version_store/_store.py:487]: Am I right in saying:

- We are just using `is_monotonic_increasing`/`is_monotonic_decreasing` from Pandas for this verification, we are not doing any of our own checks if Pandas does not know?
- We will not reorder unsorted data, it is only a check?

If so, I think it is worth being explicit in the docstring about both of these, as well as specifying what exception will be thrown if this flag is true but unsorted data is provided.

File: cpp/arcticdb/version/version_store_api.hpp
Comment (alexowens90) [inline cpp/arcticdb/version/version_store_api.hpp:99]: Am I right in assuming there's no `ensure_sorted` flag for `update` because it is effectively always true?

File: python/tests/unit/arcticdb/version_store/test_append.py
Comment (alexowens90) [inline python/tests/unit/arcticdb/version_store/test_append.py:300]: nit: `np.arange` produces data like this more concisely

File: python/tests/unit/arcticdb/version_store/test_update.py
Comment (alexowens90) [inline python/tests/unit/arcticdb/version_store/test_update.py:544]: Why is this a new test? `update` has always failed with non-timestamp indexed data hasn't it?

File: python/tests/unit/arcticdb/version_store/test_update.py
Comment (alexowens90) [inline python/tests/unit/arcticdb/version_store/test_update.py:474]: I think we also need a test that writes some `UNSORTED` data, and then calls `update` with some sorted data, as this should fail.

File: python/tests/unit/arcticdb/version_store/test_append.py
Comment (alexowens90) [inline python/tests/unit/arcticdb/version_store/test_append.py:294]: I think we also need a test that writes some `UNSORTED` data and then calls `append` with `ensure_sorted=True`, as this should fail.

File: cpp/arcticdb/entity/types.hpp
Comment (alexowens90) [inline cpp/arcticdb/entity/types.hpp:665]: Use a switch statement

File: python/arcticdb/version_store/_store.py
Comment (alexowens90) [inline python/arcticdb/version_store/_store.py:450]: Is `ensure_sorted` the best name for this? If `is_monotonic_decreasing` is `True` on the input data, it's definitely sorted, but we will reject anything that isn't monotonically _increasing_

**Review by alexowens90 — CHANGES_REQUESTED** [review]

File: cpp/arcticdb/version/version_core.cpp
Comment (qc00) [inline cpp/arcticdb/version/version_core.cpp:293]: The `TimeseriesIndex` is already checked on line 298 below. You might want to move that check up.

File: cpp/arcticdb/version/version_core.cpp
Comment (qc00) [inline cpp/arcticdb/version/version_core.cpp:541]: The index type check is already preformed above.

You might also want to extract `index_segment_reader.tsd().stream_descriptor().sorted()` out to a variable.

File: python/arcticdb/version_store/_normalization.py
Comment (qc00) [reply python/arcticdb/version_store/_normalization.py:868]: `UNKNOWN` is probably better considering the deserialized data could have a sorted index.

Comment (GemDot) [reply python/arcticdb/version_store/_normalization.py:868]: I will leave as is now, since assuming unsorted could cause unnecessary exceptions, in case we deal with the scenario that Qi commented

File: python/arcticdb/version_store/_store.py
Comment (GemDot) [reply python/arcticdb/version_store/_store.py:487]: yes. you are right with both. 

For your suggestion: regarding your second point I thing we already are explicit on that in the docstring.  Your first point is fair to add in the doctoring in a similar way we are already doing in the exception message.

File: cpp/arcticdb/version/version_store_api.hpp
Comment (GemDot) [reply cpp/arcticdb/version/version_store_api.hpp:99]: correct. As the error message says:  When calling update on sorted data, the new data must be sorted. ArcticDB relies on Pandas to detect if data is sorted - you can call DataFrame.index.is_monotonic_increasing on your input DataFrame to see if Pandas believes the data to be sorted.

There is no point to update otherwise considering our logic

File: python/tests/unit/arcticdb/version_store/test_update.py
Comment (GemDot) [reply python/tests/unit/arcticdb/version_store/test_update.py:544]: yes it is correct. Matt asked for this test If I don't miss remember

File: python/tests/unit/arcticdb/version_store/test_update.py
Comment (GemDot) [reply python/tests/unit/arcticdb/version_store/test_update.py:474]: Initially we were doing it, but after some discussions with Matt we moved towards explicitly check only the new data, not the existing one. I will leave this comment open, so we remember to discuss it next week

File: python/tests/unit/arcticdb/version_store/test_append.py
Comment (GemDot) [reply python/tests/unit/arcticdb/version_store/test_append.py:294]: Idem to previous comment

File: cpp/arcticdb/version/version_core.cpp
Comment (GemDot) [reply cpp/arcticdb/version/version_core.cpp:109]: I will explain it there instead. It was (again) after deeply discussing with Matt

File: cpp/arcticdb/version/local_versioned_engine.cpp
Comment (GemDot) [reply cpp/arcticdb/version/local_versioned_engine.cpp:374]: As we commented 2 weeks ago, by now we will leave as is

File: python/arcticdb/version_store/_store.py
Comment (GemDot) [reply python/arcticdb/version_store/_store.py:450]: Correct, but Arcticc point of view does not support currently the decreasing approach. So, as its eyes, it is not sorted. But I will leave it open to discussion for next week

File: cpp/arcticdb/version/version_core.cpp
Comment (GemDot) [reply cpp/arcticdb/version/version_core.cpp:108]: As commented above. We have to agree all together. We already discussed the names with Matt. Otherwise it is a back and forth in the PR reviews, which is not efficient.

File: cpp/arcticdb/version/version_core.cpp
Comment (GemDot) [reply cpp/arcticdb/version/version_core.cpp:293]: I disagree here. We do that check much later and only under a specific type of the row filter. On the other hand, by doing this ordering check earlier, we save unnecessary computation, in case the update is not permitted because unsorted data

File: cpp/arcticdb/version/version_core.cpp
Comment (GemDot) [reply cpp/arcticdb/version/version_core.cpp:541]: Certainly, it could be factored out and save some extra check. However, the impact of this saving would be marginal compared, at expenses of losing clarity in the code to understand what exactly we are checking here, which I think it is  important. On the other hand, we might want to change the definition of some of the checks in the future, and it could become a problem.

File: python/arcticdb/version_store/_normalization.py
Comment (alexowens90) [reply python/arcticdb/version_store/_normalization.py:870]: We can't append or update pickled data, so if that's all this enum is for I still think it should be `UNSORTED`, but not a hill I'm willing to die on.

File: python/tests/unit/arcticdb/version_store/test_update.py
Comment (mehertz) [reply python/tests/unit/arcticdb/version_store/test_update.py:440]: `update` should fail regardless of whether the data is sorted or not. Agree this would be a good descriptive test, but given the contents of the data doesn't matter this needn't block the PR IMO.

File: python/tests/unit/arcticdb/version_store/test_append.py
Comment (mehertz) [reply python/tests/unit/arcticdb/version_store/test_append.py:286]: Is this identical to the comment about update? Update is illogical. Append isn't.

File: cpp/arcticdb/version/version_core.cpp
Comment (mehertz) [reply cpp/arcticdb/version/version_core.cpp:109]: Change it.

File: cpp/arcticdb/version/version_core.cpp
Comment (mehertz) [reply cpp/arcticdb/version/version_core.cpp:108]: I'm not concerned about the naming honestly. I think this parameter is code for `sorted as far as ArcticDB is concerned`, and given this may change in the future (we should support descending data) I don't think we should be too fine-grained with the naming. If we name it `is_monotonic_ascending` we'd have to complicate matters once it supports descending data.

Comment (qc00) [reply cpp/arcticdb/version/version_core.cpp:108]: But we cannot change what `ensure_sorted` means once this is merged as it's in the public Python API.

If we ever want to support querying descending data, we have to introduce yet another conflicting argument to the Python...

File: python/tests/unit/arcticdb/version_store/test_update.py
Comment (mehertz) [reply python/tests/unit/arcticdb/version_store/test_update.py:440]: Spoke with @GemDot. Code doesn't currently reject updates if the existing data is unsorted. Modifying it to do that and adding a test.

File: cpp/arcticdb/version/version_core.cpp
Comment (qc00) [reply cpp/arcticdb/version/version_core.cpp:294]: Under this change, if the user passes a non-Timeseries argument to `update`, s/he will get an irrelevant error message about sorting instead of the correct error message that the argument type must be Timeseries indexed.

File: python/arcticdb/version_store/_store.py
Comment (GemDot) [reply python/arcticdb/version_store/_store.py:451]: validate_index_ordering

Comment (GemDot) [reply python/arcticdb/version_store/_store.py:451]: validate_index

File: cpp/arcticdb/version/version_core.cpp
Comment (GemDot) [reply cpp/arcticdb/version/version_core.cpp:109]: ok

File: cpp/arcticdb/version/version_core.cpp
Comment (GemDot) [reply cpp/arcticdb/version/version_core.cpp:294]: move 299 above the new check

File: cpp/arcticdb/version/version_core.cpp
Comment (mehertz) [reply cpp/arcticdb/version/version_core.cpp:294]: This should have an error code/category.

File: cpp/arcticdb/version/version_core.cpp
Comment (mehertz) [inline cpp/arcticdb/version/version_core.cpp:296]: `When calling update on sorted data, the input data must be sorted.`

File: python/arcticdb/version_store/_store.py
Comment (mehertz) [inline python/arcticdb/version_store/_store.py:628]:
```
If True, will verify that resulting symbol will support date range searches and update operations. This in effect tests that the previous version of the data and `data` are both sorted in ascending order. 
```

File: python/arcticdb/version_store/_store.py
Comment (mehertz) [inline python/arcticdb/version_store/_store.py:1017]: Same comment as write. Mention that it applies to each entry in the batch.

File: python/arcticdb/version_store/_store.py
Comment (mehertz) [inline python/arcticdb/version_store/_store.py:1147]: Same comment as batch_write

File: python/arcticdb/version_store/library.py
Comment (mehertz) [inline python/arcticdb/version_store/library.py:301]: Same comment as in `_store.py`.

**Review by mehertz — CHANGES_REQUESTED** [review]

File: cpp/arcticdb/version/version_core.cpp
Comment (GemDot) [reply cpp/arcticdb/version/version_core.cpp:295]: it does. category: sorting. Error code: ErrorCode. It is the same as other checks like normalization::check<ErrorCode::E_INCOMPATIBLE_INDEX>

Comment (mehertz) [reply cpp/arcticdb/version/version_core.cpp]: This is just a plain `util::check`. I think I must be missing how that would raise a sorting error?

Comment (GemDot) [reply cpp/arcticdb/version/version_core.cpp]: It is how I discussed with William and seems consistent to the rest of checks

Comment (GemDot) [reply cpp/arcticdb/version/version_core.cpp]: **util::check(index_desc.kind() == IndexDescriptor::TI**ME...  Ah sorry... this is not part of my change, it is another existing check that I just moved (as was requested by Qi). It is not related to the sort per see

**Review by GemDot — COMMENTED** [review]
Addressed all the comments

**Review by GemDot — COMMENTED** [review]
Addressed all the review comments

**Review by mehertz — APPROVED** [review]

**Pull request merged**
